source/fe/fe_pyramid_p.cc

```diff
@@ -33,87 +33,338 @@ namespace
+
+            // on the lines here are on faces in the pyramid
+            n_dofs_total_per_object[2] +=
+              fe_q.reference_cell().n_lines() * fe_q.n_dofs_per_line();
+
+            dpo.dofs_per_object_exclusive[2][1] += fe_q.n_dofs_per_line();
+            dpo.dofs_per_object_exclusive[2][2] += fe_q.n_dofs_per_line();
+            dpo.dofs_per_object_exclusive[2][3] += fe_q.n_dofs_per_line();
+            dpo.dofs_per_object_exclusive[2][4] += fe_q.n_dofs_per_line();
+            dpo.dofs_per_object_inclusive[2][1] += fe_q.n_dofs_per_line() + 2;
+            dpo.dofs_per_object_inclusive[2][2] += fe_q.n_dofs_per_line() + 2;
+            dpo.dofs_per_object_inclusive[2][3] += fe_q.n_dofs_per_line() + 2;
+            dpo.dofs_per_object_inclusive[2][4] += fe_q.n_dofs_per_line() + 2;
+
+
+            // on the faces here is on the hex on the pyramid
+            n_dofs_total_per_object[3] += fe_q.n_dofs_per_quad();
+            dpo.dofs_per_object_exclusive[3][0] += fe_q.n_dofs_per_quad();
+
+
+            n_dofs_per_object[1] += 1;
+            n_dofs_per_object[3] += fe_q.n_dofs_per_line();
+          }
+      }
+
+    unsigned int global_counter = 0;
+
+    std::vector<unsigned int> start_lines(4);
+    start_lines[0] = n_dofs_total_per_object[0] + n_dofs_per_object[0];
+    start_lines[1] = start_lines[0] + n_dofs_per_object[1];
+    start_lines[2] = start_lines[1] + n_dofs_per_object[1];
+    start_lines[3] = start_lines[2] + n_dofs_per_object[1];
+
+    dpo.object_index[1][4] = start_lines[0];
+    dpo.object_index[1][5] = start_lines[1];
+    dpo.object_index[1][6] = start_lines[2];
+    dpo.object_index[1][7] = start_lines[3];
+
+    std::vector<unsigned int> start_faces(4);
+    start_faces[0] = n_dofs_total_per_object[0] + n_dofs_total_per_object[1] +
+                     n_dofs_per_object[2];
+    start_faces[1] = start_faces[0] + n_dofs_per_object[3];
+    start_faces[2] = start_faces[1] + n_dofs_per_object[3];
+    start_faces[3] = start_faces[2] + n_dofs_per_object[3];
+
+    dpo.object_index[2][0] =
+      n_dofs_total_per_object[0] + n_dofs_total_per_object[1];
+    dpo.object_index[2][1] = start_faces[0];
+    dpo.object_index[2][2] = start_faces[1];
+    dpo.object_index[2][3] = start_faces[2];
+    dpo.object_index[2][4] = start_faces[3];
+
+
+    unsigned int start_hex = n_dofs_total_per_object[0] +
+                             n_dofs_total_per_object[1] +
+                             n_dofs_total_per_object[2];
+
+    dpo.object_index[3][0] = start_hex;
+
+    for (unsigned int current_degree = degree; current_degree > 0;
+         --current_degree)
       {
-        dpo.dofs_per_object_exclusive  = {{1}, {0}, {0, 0, 0, 0, 0}, {0}};
-        dpo.dofs_per_object_inclusive  = {{1}, {2}, {4, 3, 3, 3, 3}, {5}};
-        dpo.object_index               = {{}, {5}, {5}, {5}};
-        dpo.first_object_index_on_face = {{}, {4, 3, 3, 3, 3}, {4, 3, 3, 3, 3}};
+        FE_Q<2> fe_q(current_degree);
```

Comment:
Regarding cubic simplices: the current implementation uses evenly spaced points because it is relatively easy to describe them with barycentric polynomials. In particular, placing a support point at `(1/3, 0, 0)` is convenient because it lines up with the face centroid at `(1/3, 1/3, 0)`: it is a lot harder to make this work with general support points.

Do we have a good reference for how to do this? I found some papers which compute support points through some numerical procedure but I haven't seen an exact solution.

Reply:
I think the canonical reference for all of this kind of stuff these days is FIAT (https://fenics.readthedocs.io/_/downloads/fiat/en/latest/pdf/; for the latest version, see https://www.arxiv.org/pdf/2408.03565).

It probably doesn't matter for degree=3, but if we ever wanted to support high-order elements on triangles and other shapes, I would not be opposed if we just link in FIAT or interface with it some other way.